cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/ConsumablePartition.java

```diff
@@ -0,0 +1,118 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.data2.dataset2.lib.partitioned;
+
+import co.cask.cdap.api.common.Bytes;
+import co.cask.cdap.api.dataset.lib.Partition;
+import co.cask.cdap.api.dataset.lib.PartitionKey;
+import co.cask.cdap.proto.codec.PartitionKeyCodec;
+import com.google.common.base.Supplier;
+import com.google.common.base.Suppliers;
+import com.google.common.primitives.Ints;
+import com.google.common.primitives.Longs;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+
+import java.nio.ByteBuffer;
+
+/**
+ * Represents a {@link Partition} available for consuming.
+ */
+public final class ConsumablePartition {
+  private static final Gson GSON =
+    new GsonBuilder().registerTypeAdapter(PartitionKey.class, new PartitionKeyCodec()).create();
+
+  private final PartitionKey partitionKey;
+  private final Supplier<byte[]> partitionKeyBytes = Suppliers.memoize(new Supplier<byte[]>() {
+    @Override
+    public byte[] get() {
+      return Bytes.toBytes(GSON.toJson(partitionKey));
+    }
+  });
+
+  private ProcessState processState;
+  // the timestamp when it was marked IN_PROGRESS. it will be 0 if the ProcessState is NEW
+  private long timestamp;
+
+  public ConsumablePartition(PartitionKey partitionKey) {
+    this(partitionKey, ProcessState.NEW, 0);
+  }
+
+  public ConsumablePartition(PartitionKey partitionKey, ProcessState processState, long timestamp) {
+    this.partitionKey = partitionKey;
+    this.processState = processState;
+    this.timestamp = timestamp;
+  }
+
+  public PartitionKey getPartitionKey() {
+    return partitionKey;
+  }
+
+  public byte[] getPartitionKeyBytes() {
+    return partitionKeyBytes.get();
+  }
+
+  public long getTimestamp() {
+    return timestamp;
+  }
+
+  public ProcessState getProcessState() {
+    return processState;
+  }
+
+  public boolean isInProgress() {
+    return processState == ProcessState.IN_PROGRESS;
+  }
+
+  public void markInProgress(long timestamp) {
+    processState = ProcessState.IN_PROGRESS;
+    this.timestamp = timestamp;
+  }
+
+  public void resetState() {
+    processState = ProcessState.NEW;
+    timestamp = 0;
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+
+    ConsumablePartition that = (ConsumablePartition) o;
+
+    if (timestamp != that.timestamp) {
```

Comment:
shorter: return timestamp==that.timestamp && processState==that.processState && partitionKey.equals(that.partitionKey)